unixpwd/c/unixpwd.c

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+#include <errno.h>
+#include <malloc.h>
+#include <pwd.h>
+#include <shadow.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <limits.h>
+
+#ifdef DEVELOPMENT
+#define ETC_PASSWD   "passwd"
+#define TMP_PASSWD   "passwd.XXXXXX"
+#define ETC_SPASSWD  "shadow"
+#define TMP_SPASSWD  "shadow.XXXXXX"
+#else
+#define ETC_PASSWD   "/etc/passwd"
+#define TMP_PASSWD   "/etc/passwd.XXXXXX"
+#define ETC_SPASSWD  "/etc/shadow"
+#define TMP_SPASSWD  "/etc/shadow.XXXXXX"
+#endif
+
+#define BUFLEN 4096
+
+char           *
+unixpwd_getpwd(const char *user)
+{
+    struct passwd   pwd,
+                   *pw;
+    char            buf[BUFLEN];
+
+    if (getpwnam_r(user, &pwd, buf, BUFLEN, &pw) == 0 && pw)
+        return strdup(pw->pw_passwd);
+    if (errno == 0)
```

Comment:
The manpage says `If one wants to check errno after the call, it should be set to zero before the call.`

Reply:
You got to love C libraries. Good catch. I noticed already that you need to check both return code and the `pw` pointer.

Reply:
I've added a commit to handle this